datastore/bolt/application.go

```diff
@@ -132,240 +84,131 @@ func (a *appRepo) LoadApplicationsPagedByGroupId(ctx context.Context, gid string
 
 func (a *appRepo) SearchApplicationsByGroupId(ctx context.Context, gid string, searchParams datastore.SearchParams) ([]datastore.Application, error) {
 	var apps []datastore.Application
-	err := a.db.View(func(tx *bbolt.Tx) error {
-		b := tx.Bucket([]byte(a.bucketName))
-
-		return b.ForEach(func(k, v []byte) error {
-			var app datastore.Application
-			err := json.Unmarshal(v, &app)
-			if err != nil {
-				return err
-			}
-
-			shouldAdd := false
-			if app.GroupID == gid {
-				shouldAdd = true
-
-				if searchParams.CreatedAtStart != 0 && app.CreatedAt.Time().Unix() < searchParams.CreatedAtStart {
-					shouldAdd = false
-				}
 
-				if searchParams.CreatedAtEnd != 0 && app.CreatedAt.Time().Unix() > searchParams.CreatedAtEnd {
-					shouldAdd = false
-				}
-			}
-
-			if shouldAdd {
-				apps = append(apps, app)
-			}
+	af := &appFilter{
+		hasGroupId:   !util.IsStringEmpty(gid),
+		groupId:      gid,
+		hasStartDate: searchParams.CreatedAtStart > 0,
+		hasEndDate:   searchParams.CreatedAtEnd > 0,
+		searchParams: searchParams,
+	}
 
-			return nil
-		})
-	})
+	err := a.db.Find(&apps, a.generateQuery(af))
 
 	return apps, err
 }
 
 func (a *appRepo) FindApplicationByID(ctx context.Context, aid string) (*datastore.Application, error) {
 	var application *datastore.Application
-	err := a.db.View(func(tx *bbolt.Tx) error {
-		b := tx.Bucket([]byte(a.bucketName))
 
-		appBytes := b.Get([]byte(aid))
-		if appBytes == nil {
-			return datastore.ErrApplicationNotFound
-		}
-
-		var app *datastore.Application
-		err := json.Unmarshal(appBytes, &app)
-		if err != nil {
-			return err
-		}
-		application = app
+	err := a.db.Get(aid, &application)
 
-		return nil
-	})
+	if err != nil && errors.Is(err, badgerhold.ErrNotFound) {
+		return application, datastore.ErrApplicationNotFound
+	}
 
 	return application, err
 }
 
 func (a *appRepo) FindApplicationEndpointByID(ctx context.Context, appID string, endpointID string) (*datastore.Endpoint, error) {
 	var endpoint *datastore.Endpoint
-	err := a.db.Update(func(tx *bbolt.Tx) error {
-		b := tx.Bucket([]byte(a.bucketName))
-
-		appBytes := b.Get([]byte(appID))
-		if appBytes == nil {
-			return datastore.ErrApplicationNotFound
-		}
-
-		var app *datastore.Application
-		err := json.Unmarshal(appBytes, &app)
-		if err != nil {
-			return err
-		}
 
-		for _, v := range app.Endpoints {
-			if v.UID == endpointID {
-				endpoint = &v
+	err := a.db.ForEach(badgerhold.Where("UID").Eq(appID), func(application *datastore.Application) error {
```

Comment:
A better option would be to fetch the app using a.db.Get(app.UID) and then iterate it's endpoints looking for the one with that endpoint id